### PR TITLE
[WIP] Pure JavaScript translations

### PR DESCRIPTION
**Pull request by MatthewVita**
ex: console.log(xlj("Blue Cross Blue Shield")); // => Blått kors blå sköld

This needs some testing, but I think it's a good start!

### CODE REVIEW

**Comment by bradymiller** [comment]
Hi MatthewVita ,

Looks like you'll be loading all the translations for each language. A standard OpenEMR installation will have 7500 translations. And customized versions will have even more. For example, I know of a current group that has 80,000 of them. Won't this cause performance/memory issues on the client side?

There is also an interesting solution created by ZH Healthcare in Zend code using a function called js_xl that does a ajax call for each translation.
js_xl function is here:
interface/modules/zend_modules/public/js/application/common.js

the receiving code for the ajax call is here:
interface/modules/zend_modules/module/Application/src/Application/Controller/IndexController.php
(the ajaxZxlAction() function)

From a performance perspective, this basically migrates performance impact from huge array of translations on the client side to instead more server requests (which are not asynchronous since the info is needed before continuing processing of the script).

This is why am now a bit in between and simply loading up the specific translations prior to each script, which I learned how to do from @mdsupport . But would be nice to have a less cumbersome approach (such as yours or ZH's approach) as long as performance wasn't affected.

-brady

File: interface/main/translations/javascript_translations_provider_controller.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        }
+    }
+
+    public function get() {
+        HttpResponseHelper::send(200, generate_javascript_translation_mapping(), 'TEXT');
+    }
+}
+
+// Initialize self (note this is here until we use Zend Framework)
+$javaScriptTranslationProviderController = new JavaScriptTranslationProviderController();
```
Comment (bradymiller) [inline interface/main/translations/javascript_translations_provider_controller.php]: I am still wrapping my brain around the class and code in the same script. Don't worry, my brain will get there at some point. As an aside, I assume these types of classes can't be autoloaded in the future?

File: interface/main/translations/javascript_translations_provider_controller.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * along with this program. If not, see <http://opensource.org/licenses/gpl-license.php>;.
+ *
+ * @package OpenEMR
+ * @author  Matthew Vita <matthewvita48@gmail.com>
+ * @link    http://www.open-emr.org
+ */
+
+$fake_register_globals=false;
+$sanitize_all_escapes=true;
+$ignoreAuth=true;
```
Comment (bradymiller) [inline interface/main/translations/javascript_translations_provider_controller.php]: Will this ignoreAuth flag be required. Note we only use it for code that will be used when users are not yet authorized (such as the login page or your registration code). This flag basically means anybody can run it without being authorized.

File: interface/main/utils/http_response_helper.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://opensource.org/licenses/gpl-license.php>;.
+ *
+ * @package OpenEMR
+ * @author  Matthew Vita <matthewvita48@gmail.com>
+ * @link    http://www.open-emr.org
+ */
+
+$fake_register_globals=false;
+$sanitize_all_escapes=true;
```
Comment (bradymiller) [inline interface/main/utils/http_response_helper.php]: Above flags are not needed here. Only used before calling the globals.php, which is where the flags are used.

**Comment by MatthewVita** [comment]
Hi Brady,

Thanks for the feedback. A few things (sorry for the random list of responses... I am in a hurry this AM).

1) The solution you provided from ZH is actually kind of risky (using async false is an anti-pattern, non-performant, and is in the process of being removed from browsers: https://xhr.spec.whatwg.org/#synchronous-flag)

2) The code should only return the translations for the user's provided language. I may have screwed up the query though :)

3) I switched the language to language id of 2 (which I believe is Swedish) and the resulting array was ~200KB, which is smaller than a compressed Ember.js (~400KB). I wonder what the maximum file size of a given language would be? ~800KB?

4) As far as performance goes, the lookup is fast because the translation constants are the keys. I would like to move the script into it's own file and also do caching on the server side. We can only store 

5) The solution you are using now is clever, but I don't know if it's standard/intuitive. I'm not "married" to any particular solution. I think we need to continue our discussion/evaluate performance and memory implications before committing to any one solution.

6) The whole self-instantiation of the controller class is because we need to move to a framework that will construct controllers for us. I prefer to use classes for everything so that's why I'm invoking it automatically... ideally we would have Zend do that for us when we switch over. Sorry for the confusion.

Thanks,

Matthew

File: interface/main/tabs/main.php [hunk 4]
```diff
@@ -90,6 +90,18 @@ function isEncounterLocked( encounterId ) {
 	<?php } ?>
 }
 var webroot_url="<?php echo $web_root; ?>";
+
+// TODO: This needs to be in a <script src= ... file so that the big string doesn't make
+// the source file so wide. Not a huge deal, but would be nice.
+function xlj(key) {
+  var _translations=<?php require_once($GLOBALS['fileroot'] . "/interface/main/translations/javascript_translations_provider_controller.php"); ?>;
```
Comment (sunsetsystems) [inline interface/main/tabs/main.php]: This seems awfully convoluted and hard to follow.  Couldn't you just call generate_javascript_translation_mapping() right here?

**Comment by sunsetsystems** [comment]
Matthew could you articulate the problem that you're solving here?  I agree that it's not good to load all possible translations into the browser, even for just one language.

There is the case where a .js file has messages that need translation, and since it's a .js file PHP is not available.  Only a small subset of messages is needed for that.  Alternatively the .js file could be converted to a .php file.

**Comment by mdsupport** [comment]
1. Agree with Matthew that Zend has a more comprehensive solution ([zend-i18n](https://docs.zendframework.com/zend-i18n/translation/)) for multi-language solutions as expected of any enterprise-grade framework.  Is it possible to use that now?
2. Nothing in the translations is private.  If size permits, it may be interesting to have architecture that caches all translations upon login (limited to choice of the language) with local translations.
3. Key change that is needed in the application is to switch from using entire text as key.  All enterprise applications use a short message key to allow creation of semantic translations specific to language grammar.   That will be a big change from rather simplistic current approach of translating text fragments resulting in strange translated messages.

**Comment by sunsetsystems** [comment]
@mdsupport I'd say that discussion is beyond the scope of this PR.  We are certainly open to reworking the translation methods but tend to work at a more detailed level of discussion than "enterprise-grade".  In my experience even enterprises are capable of screwing up an IT project. :)

**Comment by sunsetsystems** [comment]
Oh and regarding your #3 we do already have support for context-sensitive translations.

**Comment by MatthewVita** [comment]
Hi Rod,

Thanks for your responses. I'll reply to your concerns in the sections below.

## Design

You were wondering why I have `Tabs Main` calling `JavaScriptTranslationsProviderController` that calls the `Translation` layer (which I'm treating as the service layer) when I could have just called the translations method right from the view. In my [work-in-progress pull request to add OpenEMR Product Registration](https://github.com/openemr/openemr/pull/257), I introduced some modern architecture that appears to be well-received and will be embraced in the [modernization project](http://open-emr.org/wiki/index.php/Codebase_Refactor). It uses a clear separation of the client (JavaScript driven) and the server (JSON API), both using the MVC design pattern and are unit-testable. I followed a similar route in this PR.

A major goal of mine (outside of helping this project wherever I can) is to start conversations (via PRs) in line with the modernization efforts to see if we can clean up code (i.e.: use MVC, standardize conventions, write testable code, remove db calls in views, separate client and server responsibilities, and use dynamic JavaScript frameworks/libraries, etc). I want to add value to the project and I want to find the right amount of strictness in terms of practical software engineering patterns that work for everyone moving forward.

## PHP-Driven JS Issue

One item that Robert brought up with the modernization efforts is unit testing. I tend to agree that tests are important (have saved my butt countless times...). The JavaScript I introduced in my other PR is not unit testable with [Karma](https://karma-runner.github.io/1.0/index.html) and [Jasmine](http://jasmine.github.io/) because of the fact it is not valid JavaScript with the PHP rendering.

## Client Side Translation Usage

The JavaScript I am writing in the other PR is using translations and since the modernization efforts appear to favor React.js and Knockout.js, there may be a need for a decent client side translation ability. I realize that ZH provided a JS solution for this via the `js_xl` method, but that is actually kind of risky as the code under the hood will be rendered broken once [`async: false` is removed from browsers.](https://xhr.spec.whatwg.org/#synchronous-flag).

Brady's solution of passing the translations as needed to the "pure" js files is clever. I'm not 100% happy with it, but it's a working solution and I'd be fine moving forward with it as long as it is well document (I'll volunteer to document it). With these things, consistency is the most important thing really.

## Performance/Memory Implications

I've been on 2 enterprise healthcare projects that use Angular.js as the client with the i18n all stored on the client. Perfomance and memory were not noticably impacted.

[Angular Translate](https://github.com/angular-translate/angular-translate/tree/master/demo), [Ember i18n](https://github.com/jamesarosen/ember-i18n), and [React Translate Component](https://github.com/martinandert/react-translate-component) all do this and I'm sure (based on each having hundreds of Github stars and the sheer popularity of the frameworks) they are used in production.

My tests with the OpenEMR Swedish language translations ended up weighing in around 200KB. This is a small footprint compared to bringing in a decent size JavaScript library or framework (for example, uncompressed jQuery.js is ~200KB and compressed Ember.js is around ~400KB). Even if a particular language translation ends up weighing in around 1MB, that's really not much of a hit if we cache and separate the dictionary in it's own js file. However, if that _is_ a problem than I understand and feel that Brady's solution will do (as long as it is well documented and used consistency moving forward)

Sorry for the wall of text. I look forward to finding the right balance of engineering rigor and contributing more to the project!

-m

**Comment by robertdown** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

Matthew,

Glad to see such a strong push to move to the more modern project, it's
too big for me to take on by myself.

Due to the scale of the project I think we need to start in a very
specific place. Here are my thoughts

We need to abstract the DB interaction away from *_all views_*. This
likely will take place in a couple of parts.

I suggest we first explore ADODB and how to properly extend it (read:
subclass) to meet our needs. This should be created as a model.

- From there I suggest we move the logic of a view page to the top the
PHP page and introduce the new OO-based interactions. This allows us to
move the business logic out of the view without having to create a full
model for each component.

Second, I think we should examine the use of a PHP templating engine. I
see Smarty used sporadically used in the project, but have not used it
much myself. I have a little bit of experience with Twig and personally
enjoy it. It is made by Symfony.

Regardless of _what_ we choose, we should definitely land on something
and get all of these `<?php echo... ?>` out of the code.

Next, I think we can begin to discuss how we want to hash out the
service layer... which is a conversation for another day.

-----BEGIN PGP SIGNATURE-----

iQIcBAEBCAAGBQJX7cmTAAoJEPI0tmejVt7tiTkP/3cbepoj1i7ChtETHPML4eJD
Gn4B6ExccRFcjn6Yz0Qdqd5IiFddRH+YxBH4N878UyJhpP98DK1BOFMqxheouw3H
vjZBUuTsQG5Dg0sN21BxTWNQ/L1bqWEIzsPsRkoVYAeVF/HFZC5TpBP3k+YCQmlI
tAsxlWvaRiZenxbTyZMLczyGLkqaisidFEMlxeP7gLPZhO6wx/3OWwHjHVWFer0E
/vO9v0FBXDmgTsLEcIPiFxLsicT0thpR9v3nGQCHdWO0Ix4sKrP8QdPSD0GcX1r9
UZ9a4N7XXYoC86rJ8T5TQb6Kfgdfyu/gGEfdd9J6+qH8XJS3SKMzkzx8KJmLfvLW
dkb0zkV2qKa2E/G6nZe4w+MDsVJT6jUKDC7PzKITTY4/xSHHN06jN5Zn+iha1gDN
8JGiVO19T9SISjZP7XKwGK0o+hjHxvG/L0CB9kYpMRxQzwN1gN1kTvKPjtgczIj5
lssnCU5KitjWEqpb803Dq0HiBHkc10LbPk11mNMh+g37I8dNopwNCB1/m+vTKwav
+OBpsyEVb+BEr8ikvOyuj+OFEO+B9i+uqbtQVj6WwfP6rmmXQ4NGlwvTQsDoBxca
Qffv4ik+xDuB8NIVrcHIQpTfNz8KFEU0U+zZEW6U6pyRPIqtf+NFRC8ATQhs5GEZ
h7sbc/7SvW4/L4tGLCp+
=rVtj
-----END PGP SIGNATURE-----

**Comment by Wakie87** [comment]
@robertdown  I am currently in the process of removing ADODB and moving to PDO. You can see progress here:
https://github.com/Wakie87/openemr/tree/database

**Comment by robertdown** [comment]
Awesome, love it!

**Comment by sunsetsystems** [comment]
Matthew, thanks for the insight.  I will be looking more closely at the methods proposed.

I do agree with Robert that abstracting database interaction is the sensible first step, and that we should beware of biting off more than we can chew at one time.  We should also beware of complexity that discourages contributions from other developers.

Some sort of translation caching seems right to me, loading into the browser only what's needed as needed.  With a cache this should not be a performance hardship.  The number of entries in lang_definitions for a language can be huge and the number actually used in a given session will be much smaller.

**Comment by MatthewVita** [comment]
@Rod @Brady: Sounds good to me. I will take the time to document Brady's solution of passing the translations as needed to the "pure" js files and close out this PR. Will follow up soon!

I am also going to update the modernization project page with a task to replace the risky `js_xl` JavaScript method that I mentioned before.

@Robert: I agree the DB layer can be our first item to tackle and ADODB seems to be similar to other ORMs I've used. Do you have a branch started or should I create one?

I agree that the project would benefit from a modern template engine. Also feel that using Knockout.js more for the UI would be a big improvement.

Thanks,
Matthew

**Comment by MatthewVita** [comment]
NOTE: I've update the documentation with an example of how to use the translations to JS pattern: http://www.open-emr.org/wiki/index.php/Development_Policies#Internationalization

**Comment by robertdown** [comment]
@sunsetsystems Rod, this may be the first time we've agreed on the modernization project--Woo-hoo! 

@MatthewVita See @Wakie87 comment, he's moving everything to PDO (Which is awesome).

**Comment by bradymiller** [comment]
@MatthewVita ,
Thanks for updating the wiki. Do you think it makes sense to create a separate php file (that builds the json translation array) for each js script that needs these definitions (perhaps store in library/js/translations with same name as js file but with a php at the end of it) to keep it standardized. Then would simply include it before the js file include. Then as the definition library grows or use lots more client side code (ie. knockout), could think about expanding the solution.

**Comment by MatthewVita** [comment]
@bradymiller,

I think we'll get mileage with the pattern that you introduced/that was recently documented. Let's just run with it for now.

Thanks,
Matthew

**Comment by MatthewVita** [comment]
Also, @Wakie87 and @robertdown do you feel that PDO integration is preferred over an ORM? Can these be combined?

_disclosure: I'm biased because I prefer ORMs_
